File: fsharp/robot-name/RobotName.fs
module RobotName

open System
open System.Collections.Generic

type Robot = {
    runningNumber: int
    name: string }

let private rnd: Random = new Random()
let mutable private numRobots = 0
let private givenIDs = new HashSet<string>()

let getRandomChar() = 
    rnd.Next(65, 91) |> char

let getRandomNum() =
    rnd.Next(48, 58) |> char

let getRandomID() =
    let randomChars = [| getRandomChar(); getRandomChar(); getRandomNum(); getRandomNum(); getRandomNum() |]
    new string(randomChars)

let mkRobot() = 
    let mutable id = getRandomID()
    numRobots <- numRobots + 1
    while givenIDs.Contains(id) do
        id <- getRandomID()
    givenIDs.Add(id) |> ignore
    { runningNumber = numRobots; name = id }

let name (robot: Robot) = robot.name

let reset (robot: Robot) = 
    let mutable newid = getRandomID()
    while givenIDs.Contains(newid) do
        newid <- getRandomID()
    givenIDs.Add(newid) |> ignore
    {robot with name = newid}

File: csharp/robot-name/RobotName.cs
using System;
using System.Collections.Generic;

public class Robot
{
    private static readonly Random rnd = new Random();
    // although "readonly" items can be added
    // must be static in order to be seen by ALL robots
    private static readonly HashSet<string> given = new HashSet<string>();
    public string Name { get; private set; }
    public Robot()
    {
        Reset();
    }
    public void Reset()
    {
        bool done = false;
        string name = String.Empty;
        while (!done)
        {
            char firstLetter = (char)rnd.Next(65, 91);
            char secondLetter = (char)rnd.Next(65, 91);
            char firstDigit = (char)rnd.Next(48, 58);
            char secondDigit = (char)rnd.Next(48, 58);
            char thirdDigit = (char)rnd.Next(48, 58);
            name = new string(new char[] { firstLetter, secondLetter, firstDigit, secondDigit, thirdDigit });
            if (given.Add(name))
            {
                done = true;
            }
        }
        Name = name;
    }
}